src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -86,87 +87,102 @@
    pool :- jruby-schemas/pool-queue-type]
   (= expected-pool-size (count (.getRegisteredElements pool))))
 
-(schema/defn ^:always-validate
-  swap-and-drain-pool!
-  "Replace the current pool with a new pool and drain the old pool,
-  optionally refilling the new pool with fresh jrubies."
+(schema/defn collect-all-jrubies :- [JRubyInstance]
+  "Locks the pool and borrows all the instances"
+  [pool-state :- jruby-schemas/PoolState]
+  (let [pool-size (:size pool-state)
+        pool (:pool pool-state)
+        borrow-fn (fn [_] (jruby-internal/borrow-from-pool!*
+                           jruby-internal/borrow-without-timeout-fn
+                           pool))]
```

Comment:
I think that instead of passing in the `pool-state` here, you could pass in the `pool-context` and then use `jruby-internal/get-pool-size` and `jruby-internal/borrow-from-pool`, both of which take the `pool-context`. That way you don't have to basically recreate `borrow-from-pool` here.